python/service/service.py

```diff
@@ -188,6 +188,71 @@ def remove_device_from_pairing_list(self, device_id):
                 )
                 del self.paired_devices[device_id]
 
+    def handle_method_invoke(self, device_data, event):
+        body = event.body_as_json()
+        thief = body.get(Fields.THIEF, {})
+        method_guid = thief.get(Fields.SERVICE_ACK_ID)
+        method_name = thief.get(Fields.METHOD_NAME)
+
+        logger.info("------------------------------------------------------------------")
+        logger.info("received method invoke method={}, guid={}".format(method_name, method_guid))
+
+        request = CloudToDeviceMethod(
+            method_name=method_name,
+            payload=thief.get(Fields.METHOD_INVOKE_PAYLOAD, None),
+            response_timeout_in_seconds=thief.get(
+                Fields.METHOD_INVOKE_RESPONSE_TIMEOUT_IN_SECONDS, None
+            ),
+            connect_timeout_in_seconds=thief.get(
+                Fields.METHOD_INVOKE_CONNECT_TIMEOUT_IN_SECONDS, None
+            ),
+        )
+
+        # TODO: don't hold this lock while calling into registry manager
+        # TODO: wrap registry manager call in try/except to return result to device
+        logger.info("------------------------------------------------------------------")
+        logger.info("invoking {}".format(method_guid))
+        try:
+            with self.registry_manager_lock:
```

Comment:
Why do we need a lock for the registry manager client?

Reply:
I don't like how I did this.  It's because I need to re-create the object when the token expires, and I destroy the old one in a fiarly nasty way (deleting internal objects and other things that make me feel dirty.).  I need to fix this.